compiler/src/dotty/tools/dotc/typer/Inliner.scala

```diff
@@ -149,12 +149,26 @@ object Inliner {
 
   /** Replace `Inlined` node by a block that contains its bindings and expansion */
   def dropInlined(inlined: tpd.Inlined)(implicit ctx: Context): Tree = {
-    val reposition = new TreeMap {
-      override def transform(tree: Tree)(implicit ctx: Context): Tree = {
-        super.transform(tree).withPos(inlined.call.pos)
+    if (enclosingInlineds.nonEmpty) inlined // remove in the outer inlined call
+    else {
+      val reposition = new TreeMap {
+        private[this] var pos: Position = inlined.pos
+        override def transform(tree: Tree)(implicit ctx: Context): Tree = {
+          tree match {
+            case inlined: Inlined =>
+              val last = pos
+              pos = inlined.call.pos
+              val res = tpd.seq(inlined.bindings.map(transform), transform(inlined.expansion)).withPos(last)
+              pos = last
+              res
+            case tree =>
+              if (pos.exists) super.transform(tree).withPos(pos)
```

Comment:
@nicolasstucki says `!pos.exists` is only possible if `inlined.call` is `emptyTree`, which means that this is not a *call* to an inline method but an *argument* to it — and then `bindings` should be empty.
We should update docs for `Inlined` about this.

Reply:
This does not look like the right algorithm. We have enclosingInlineds, which is interpreted as follows:

 - An Inlined node with a non-empty call is a plain inlined node
 - An inlined node with an empty tree as call in an inlined argument which _cancels_ the enclosing inlined call.

So one should start with the outermost InlinedCall and update all positions in its expansion to the call position, except in those regions where there are as many empty as non-empty calls in `enclosingInlineds`.

Reply:
That sounds like the algorithm I implemented though I use the pos state to know if it has been cancelled or not. I will change it to use `enclosingInlineds` directly.